sw-emulator/lib/periph/src/ml_dsa87.rs

```diff
@@ -0,0 +1,1254 @@
+            name1: ReadOnlyRegister::new(Self::NAME1_VAL),
+            version0: ReadOnlyRegister::new(Self::VERSION0_VAL),
+            version1: ReadOnlyRegister::new(Self::VERSION1_VAL),
+            control: ReadWriteRegister::new(0),
+            status: ReadOnlyRegister::new(Status::READY::SET.value),
+            iv: Default::default(),
+            seed: Default::default(),
+            sign_rnd: Default::default(),
+            message: Default::default(),
+            verification_result: Default::default(),
+            sk_out: [0; 1224],
+            sk_in: [0; 1224],
+            pk: [0; 648],
+            signature: [0; 1157],
+            timer: Timer::new(clock),
+            op_complete_action: None,
+        }
+    }
+
+    fn read_access_fault(&self, _size: RvSize, _index: usize) -> Result<RvData, BusError> {
+        Err(BusError::LoadAccessFault)
+    }
+
+    fn write_access_fault(
+        &self,
+        _size: RvSize,
+        _index: usize,
+        _val: RvData,
+    ) -> Result<(), BusError> {
+        Err(BusError::StoreAccessFault)
+    }
+
+    // Clear registers
+    fn zeroize(&mut self) {}
+
+    /// On Write callback for `control` register
+    ///
+    /// # Arguments
+    ///
+    /// * `size` - Size of the write
+    /// * `val` - Data to write
+    ///
+    /// # Error
+    ///
+    /// * `BusError` - Exception with cause `BusError::StoreAccessFault` or `BusError::StoreAddrMisaligned`
+    pub fn on_write_control(&mut self, size: RvSize, val: RvData) -> Result<(), BusError> {
+        // Writes have to be Word aligned
+        if size != RvSize::Word {
+            Err(BusError::StoreAccessFault)?
+        }
+
+        // Set the control register
+        self.control.reg.set(val);
+
+        match self.control.reg.read_as_enum(Control::CTRL) {
+            Some(Control::CTRL::Value::GEN_KEY)
+            | Some(Control::CTRL::Value::SIGN)
+            | Some(Control::CTRL::Value::VERIFY) => {
+                // Reset the Ready and Valid status bits
+                self.status
+                    .reg
+                    .modify(Status::READY::CLEAR + Status::VALID::CLEAR);
+
+                self.op_complete_action = Some(self.timer.schedule_poll_in(ML_DSA87_OP_TICKS));
+            }
+            _ => {}
+        }
+
+        if self.control.reg.is_set(Control::ZEROIZE) {
+            self.zeroize();
+        }
+
+        Ok(())
+    }
+
+    fn gen_key(&mut self) {
+        let seed_bytes = &bytes_from_words_le(&self.seed);
+        let mut rng = StdRng::from_seed(*seed_bytes);
+        let (pk, sk) = try_keygen_with_rng(&mut rng).unwrap();
+
+        self.pk = words_from_bytes_le(&pk.into_bytes());
+        self.sk_out = words_from_bytes_le(&sk.into_bytes());
+    }
+
+    fn sign(&mut self) {
+        let seed_bytes = &bytes_from_words_le(&self.seed);
+        let mut rng = StdRng::from_seed(*seed_bytes);
+
+        let secret_key_bytes = &bytes_from_words_le(&self.sk_in);
```

Comment:
I thought the peripheral was supposed to regen the key from a seed on every signing operation? I think the plan is that KV slots will only hold the seed and HW will generate the private key from the seed as-needed.

Reply:
I think what you say is indeed the actual use case. I think the dilithium reflects the ecc384 engine which can also take a key as input, which I was told was for early validation/ testing.